src/object/cli_ec.c

```diff
@@ -999,6 +1024,53 @@ obj_reasb_req_dump(struct obj_reasb_req *reasb_req, d_sg_list_t *usgl,
 	D_PRINT("\n");
 }
 
+static void
+ec_recov_recx_seg_add(struct obj_reasb_req *reasb_req, daos_recx_t *recx,
+		      daos_recx_t *r_recx, uint32_t *r_idx, uint32_t *start_idx,
+		      daos_size_t iod_size, d_sg_list_t *sgl,
+		      struct obj_ec_seg_sorter *sorter)
+{
+	struct obj_ec_fail_info	*fail_info = reasb_req->orr_fail;
+	struct daos_oclass_attr	*oca = reasb_req->orr_oca;
+	uint64_t		 stripe_rec_nr = obj_ec_stripe_rec_nr(oca);
+	uint64_t		 cell_rec_nr = obj_ec_cell_rec_nr(oca);
+	uint64_t		 stripe_total_sz, cell_sz, recx_idx, recx_nr;
+	uint32_t		 i, tgt, tgt_nr, stripe_nr;
+	void			*buf_sgl, *buf_stripe, *buf;
+	d_iov_t			 iov;
+
+	D_ASSERT(fail_info != NULL);
+	D_ASSERT(sgl->sg_nr == 1);
+	D_ASSERT(recx->rx_nr % stripe_rec_nr == 0);
+
+	buf_sgl = sgl->sg_iovs[0].iov_buf;
+	cell_sz = obj_ec_cell_rec_nr(oca) * iod_size;
+	stripe_total_sz = cell_sz * obj_ec_tgt_nr(oca);
+	stripe_nr = recx->rx_nr / stripe_rec_nr;
+	recx_nr = recx->rx_nr / obj_ec_data_tgt_nr(oca);
+	for (tgt = 0, tgt_nr = 0; tgt < obj_ec_tgt_nr(oca); tgt++) {
+		if (obj_ec_tgt_in_err(fail_info->efi_tgt_list,
+				      fail_info->efi_ntgts, tgt))
+			continue;
```

Comment:
so if it reach the maxim fault toleration, it does not need to recovery anymore?

Reply:
yes for now, and if #tgts_failed > limit, then it is not recoverable.
But probably some recoverable cases can be refined later, for example if some tgts in rebuilding, in that case maybe can return EAGAIN?